File: android/src/main/kotlin/net/tlserver6y/flutter_rotation_sensor/FlutterRotationSensorPlugin.kt
package net.tlserver6y.flutter_rotation_sensor

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink
import io.flutter.plugin.common.EventChannel.StreamHandler
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class FlutterRotationSensorPlugin : FlutterPlugin, MethodCallHandler, SensorEventListener,
  StreamHandler {
  private lateinit var methodChannel: MethodChannel
  private lateinit var eventChannel: EventChannel
  private lateinit var sensorManager: SensorManager
  private var eventSink: EventSink? = null
  private var sensor: Sensor? = null
  private var samplingPeriod = 200000
  private var sensorType = Sensor.TYPE_ROTATION_VECTOR

  // === FlutterPlugin ===

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    val context = flutterPluginBinding.applicationContext
    sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
    setupMethodChannel(flutterPluginBinding.binaryMessenger)
    setupEventChannel(flutterPluginBinding.binaryMessenger)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    teardownMethodChannel()
    teardownEventChannels()
  }

  private fun setupMethodChannel(messenger: BinaryMessenger) {
    methodChannel = MethodChannel(messenger, "rotation_sensor/method")
    methodChannel.setMethodCallHandler(this)
  }

  private fun teardownMethodChannel() {
    methodChannel.setMethodCallHandler(null);
  }

  private fun setupEventChannel(messenger: BinaryMessenger) {
    eventChannel = EventChannel(messenger, "rotation_sensor/orientation")
    eventChannel.setStreamHandler(this)
  }

  private fun teardownEventChannels() {
    eventChannel.setStreamHandler(null)
    onCancel(null)
  }

  // === MethodCallHandler ===

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "setSamplingPeriod" -> {
        val samplingPeriod = call.arguments as? Int ?: run {
          return result.error(
            "INVALID_ARGUMENTS",
            "Int samplingPeriod is required",
            null
          )
        }
        setSamplingPeriod(samplingPeriod)
        return result.success(null)
      }
      "setReferenceFrame" -> {
        val referenceFrame = call.arguments as? String ?: run {
          return result.error(
            "INVALID_ARGUMENTS",
            "String referenceFrame is required",
            null
          )
        }
        setReferenceFrame(referenceFrame)
        result.success(null);
      }
      else -> result.notImplemented()
    }
  }

  private fun setSamplingPeriod(samplingPeriod: Int) {
    if (samplingPeriod == this.samplingPeriod) return
    this.samplingPeriod = samplingPeriod
    resubscribe()
  }

  private fun setReferenceFrame(referenceFrame: String) {
    val sensorType = when (referenceFrame) {
      "arbitrary", "arbitraryCorrected" -> Sensor.TYPE_GAME_ROTATION_VECTOR
      "magneticNorth", "trueNorth" -> Sensor.TYPE_ROTATION_VECTOR
      else -> Sensor.TYPE_ROTATION_VECTOR
    }
    if (sensorType == this.sensorType) return
    this.sensorType = sensorType
    resubscribe()
  }

  // === SensorEventListener ===

  override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}

  override fun onSensorChanged(event: SensorEvent) {
    eventSink?.success(
      arrayListOf(
        event.values[0].toDouble(),
        event.values[1].toDouble(),
        event.values[2].toDouble(),
        event.values[3].toDouble(),
        // Estimated heading accuracy may not exist.
        event.values.getOrElse(4, { -1.0f }).toDouble(),
        event.timestamp,
      )
    )
  }

  // === StreamHandler ===

  override fun onListen(arguments: Any?, events: EventSink) {
    eventSink = events
    subscribe()
  }

  override fun onCancel(arguments: Any?) {
    eventSink = null
    unsubscribe()
  }

  private fun noListeners(): Boolean = eventSink == null

  private fun defaultSensor(): Sensor? {
    val sensor = this.sensor
    val sensorType = this.sensorType
    if (sensor?.type == sensorType) return sensor
    return sensorManager.getDefaultSensor(sensorType)?.also {
      this.sensor = it
    } ?: run {
      eventSink?.error(
        "UNAVAILABLE",
        "Sensor not found",
        "It seems that your device has no ${
          when (sensorType) {
            Sensor.TYPE_ROTATION_VECTOR -> "rotation vector sensor"
            Sensor.TYPE_GAME_ROTATION_VECTOR -> "game rotation vector sensor"
            else -> "needed sensor"
          }
        }."
      )
      null
    }
  }

  private fun subscribe() {
    defaultSensor()?.let {
      sensorManager.registerListener(this, it, samplingPeriod)
    }
  }

  private fun unsubscribe() {
    sensorManager.unregisterListener(this)
  }

  private fun resubscribe() {
    if (noListeners()) return
    unsubscribe()
    subscribe()
  }
}

File: ios/flutter_rotation_sensor/Sources/flutter_rotation_sensor/FlutterRotationSensorPlugin.swift
import Flutter
import UIKit
import CoreMotion

public class FlutterRotationSensorPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

  private var eventChannel: FlutterEventChannel
  private let motionManager: CMMotionManager
  private var eventSink: FlutterEventSink?
  private var referenceFrame: CMAttitudeReferenceFrame = .xMagneticNorthZVertical

  public static func register(with registrar: FlutterPluginRegistrar) {
    let methodChannel = FlutterMethodChannel(name: "rotation_sensor/method", binaryMessenger: registrar.messenger())
    let eventChannel = FlutterEventChannel(name: "rotation_sensor/orientation", binaryMessenger: registrar.messenger())
    let motionManager = CMMotionManager()
    motionManager.deviceMotionUpdateInterval = 0.2
    let instance = FlutterRotationSensorPlugin(eventChannel: eventChannel, motionManager: motionManager)
    registrar.addMethodCallDelegate(instance, channel: methodChannel)
    eventChannel.setStreamHandler(instance)
  }

  public init(eventChannel: FlutterEventChannel, motionManager: CMMotionManager) {
    self.eventChannel = eventChannel
    self.motionManager = motionManager
  }

  public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
    eventChannel.setStreamHandler(nil)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "setSamplingPeriod":
      guard let samplingPeriod = call.arguments as? Int32 else {
        result(FlutterError(
          code: "INVALID_ARGUMENTS",
          message: "Int32 samplingPeriod is required",
          details: nil
        ))
        return
      }
      setSamplingPeriod(samplingPeriod)
      result(nil)
    case "setReferenceFrame":
      guard let referenceFrame = call.arguments as? String else {
        result(FlutterError(
          code: "INVALID_ARGUMENTS",
          message: "String referenceFrame is required",
          details: nil
        ))
        return
      }
      setReferenceFrame(referenceFrame)
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func setSamplingPeriod(_ samplingPeriod: Int32) {
    guard samplingPeriod != Int32((motionManager.deviceMotionUpdateInterval * 1000000).rounded()) else { return }
    motionManager.deviceMotionUpdateInterval = Double(samplingPeriod) * 0.000001
  }

  private func setReferenceFrame(_ referenceFrame: String) {
    let referenceFrame: CMAttitudeReferenceFrame = switch referenceFrame {
    case "arbitrary": .xArbitraryZVertical
    case "arbitraryCorrected": .xArbitraryCorrectedZVertical
    case "magneticNorth": .xMagneticNorthZVertical
    case "trueNorth": .xTrueNorthZVertical
    default: .xMagneticNorthZVertical
    }
    guard referenceFrame != self.referenceFrame else { return }
    self.referenceFrame = referenceFrame
    resubscribe()
  }

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    subscribe()
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    unsubscribe()
    return nil
  }

  private func noListeners() -> Bool { eventSink == nil }

  private func handleDeviceMotion(motion: CMDeviceMotion?, error: Error?) {
    guard let motion else {
      if let error {
        eventSink?(FlutterError(code: "UNKNOWN", message: error.localizedDescription, details: nil))
      }
      return
    }
    let quaternion = motion.attitude.quaternion
    let rotationVector: [Any] = [
      quaternion.x,
      quaternion.y,
      quaternion.z,
      quaternion.w,
      -1.0,
      Int64((motion.timestamp * 1000000000).rounded())
    ]
    DispatchQueue.main.async {
      self.eventSink?(rotationVector)
    }
  }

  private func subscribe() {
    guard motionManager.isDeviceMotionAvailable else {
      eventSink?(FlutterError(code: "UNAVAILABLE", message: "Device motion unavailable", details: nil))
      return
    }
    motionManager.startDeviceMotionUpdates(using: referenceFrame, to: OperationQueue(), withHandler: handleDeviceMotion)
  }

  private func unsubscribe() {
    motionManager.stopDeviceMotionUpdates()
  }

  private func resubscribe() {
    if noListeners() { return }
    unsubscribe()
    subscribe()
  }
}
